### ISSUE TITLE
test_target_mapping_before_alloc.F90: Fails to compile due to con't line in preprocessor argument

### ISSUE THREAD

**Issue by tob2** [SOLLVE/sollve_vv#464]
The issue is older, even though I only spotted it now. The **testcase was added via Pull Request #306** and currently has:
https://github.com/SOLLVE/sollve_vv/blob/e51d17b0b69182619be6833685e496a2401a2cbd/tests/5.0/target/test_target_mapping_before_alloc.F90#L52-L53

The problem is the continuation line, which in Fortran uses `&` as end-of-line marker.

However, **after preprocessing,** that line is changed to the following; the problem is the `& &` which appears mid line:
```F90
    errors = errors + test_and_set_verbose(scalar .ne. 80 .OR. a(2) .ne. 2 .OR. test_struct%var .ne. 1 & & .OR. test_struct%b(2) .ne. 2, "condition", "tests/5.0/target/test_target_mapping_before_alloc.F90", 52)
```
That's at least the result with GCC's C preprocessor, including the one internally used by gfortran. That causes **gfortran to fail with**
```
tests/5.0/target/test_target_mapping_before_alloc.F90:52:104:

   52 |     OMPVV_TEST_AND_SET_VERBOSE(errors, scalar .ne. 80 .OR. a(2) .ne. 2 .OR. test_struct%var .ne. 1 &
      |                                                                                                        1
Error: Syntax error in argument list at (1)
```

For gfortran (with `-ffree-line-length-none`), the **solution** is to remove the two `&`:
```diff
--- a/tests/5.0/target/test_target_mapping_before_alloc.F90
+++ b/tests/5.0/target/test_target_mapping_before_alloc.F90
@@ -52,2 +52,2 @@ CONTAINS
-    OMPVV_TEST_AND_SET_VERBOSE(errors, scalar .ne. 80 .OR. a(2) .ne. 2 .OR. test_struct%var .ne. 1 &
-         & .OR. test_struct%b(2) .ne. 2)
+    OMPVV_TEST_AND_SET_VERBOSE(errors, scalar .ne. 80 .OR. a(2) .ne. 2 .OR. test_struct%var .ne. 1 \
+         .OR. test_struct%b(2) .ne. 2)
```

The backslash `\` is only syntactical sugar, making clear that it is a continuation line – handled by the C preprocessor. I think that improves readability – but could also be removed.

_I don't know whether there are issues with other compilers with regards to the line length (with default sollve_vv compiler flags) or preprocessing –  or whether it works with all Fortran compilers out of the box._

@spophale , @seyonglee

**Comment by seyonglee** [comment]
We have discussed a similar issue at #377. My suggestion is to use multiple print statements with short messages instead of having a single print statement with a long message with line continuation, since this is a base language issue, not related to the OpenMP.

**Comment by tob2** [comment]
> We have discussed a similar issue at #377, which was about using macro combined with line continuation.

Which is kind of the same.

_One problem is that the C preprocessor is not well defined for Fortran. In the old way (traditional mode), the preprocessing is somewhat language agnostic, but with newer features (like '##'), the preprocessor has knowledge/assumptions about the base language, such that a native C/C++ preprocessor does not work well with Fortran. – Some vendors support the newer features and created a Fortran-aware "C" preprocessor, which supports the newer C-standard preprocessing features. [gfortran hasn't and uses internally the preprocessor in the traditional mode (`cpp -traditional-cpp`).]_

In #377, there is a continuation line in a string of the form `"abc&` / `&def` – which outside CPP would be `abcdef`. But with preprocessing, as the string is not terminated before the linebreak, it becomes `"abc& &def"` with embedded `& &`. – Or as real-world example for #377, the result is: `... offloading off, it is not&                   &possible ...` (GitHub eats the spaces, there are 19 spaces between the ampersands.)
Contrary to the problem reported in the first comment, that's a valid Fortran string and readable. (Albeit, the string is not as expected.)

For the #377 file, I think one solution would be again to honor that it is preprocessed and use:
```diff
--- a/tests/4.5/target_teams_distribute_parallel_for/test_target_teams_distribute_parallel_for_if_no_modifier.F90
+++ b/tests/4.5/target_teams_distribute_parallel_for/test_target_teams_distribute_parallel_for_if_no_modifier.F90
@@ -76,2 +76,2 @@
-      OMPVV_WARNING_IF(raiseWarningDevice .eq. 0, "Initial number of threads on&
-      & device was 1. It is not possible to test the if for the parallel directive")
+      OMPVV_WARNING_IF(raiseWarningDevice .eq. 0, "Initial number of threads on" // \
+      " device was 1. It is not possible to test the if for the parallel directive")
```
(likewise for other strings). This uses `//` to concatenate the strings (at compile time) and the tailing `\` is again syntactical sugar – which is supposed to highlight that the string is preprocessed. (While for some code readers, the `\` might be just confusing.)

**Comment by tmh97** [comment]
@tob2 is using `//` followed by `\' a case that will work for all strings (at least the strings we use for error messages and MESSAGES? If so I am comfortable with this change. With that being said, a lot of our tests use the && line continuation strategy so we will probably need to make changes to all of our .F90 tests accordingly

**Comment by tmh97** [comment]
@tob2 for #pragma omp statements with continuations it means // and \ aren't compiling properly, so I'm assuming & & will still be best for those cases? (apologies for ambiguity. Thanks again for your help, always much appreciated!

**Comment by tob2** [comment]
> @tob2 is using `//` followed by `' a case that will work for all strings (at least the strings we use for error messages and MESSAGES?

I am not sure, I understand the question. `//` is Fortran's way to concatenate strings. Thus,
```F90
#define MACRO(x) f(x)
MACRO ("ABC" //
       "DEF")
```
is changed by the preprocessor to

```F90
f("ABC" //       "DEF")
```
which Fortran concats to `f("ABCDEF")` internally.

 * * *
> With that being said, a lot of our tests use the && line continuation strategy

Using `&` ... `&` _is_ the way to go if _no_ macro expansion involved. If there _is,_ the macro expansion eats the line breaks, causing problems with now mid-line ampersands.

**Comment by nolanbaker31** [comment]
Solved by PR #570, closing.
